modules/contracts/src.sol/CMCAdjudicator.sol

```diff
@@ -81,47 +87,72 @@ contract CMCAdjudicator is CMCCore, CMCAccountant, ICMCAdjudicator {
     emit ChannelDisputed(msg.sender, address(this), channelDispute);
   }
 
-  function defundChannel(CoreChannelState calldata ccs)
-    external
-    override
-    onlyViaProxy
-    nonReentrant
-    validateChannel(ccs)
-  {
+  function defundChannel(
+    CoreChannelState calldata ccs,
+    address[] calldata assetIds,
+    uint256[] calldata indices
+  ) external override onlyViaProxy nonReentrant validateChannel(ccs) {
+    // These checks are not strictly necessary, but it's a bit cleaner this way
+    require(assetIds.length > 0, "CMCAdjudicator: NO_ASSETS_GIVEN");
+    require(indices.length <= assetIds.length, "CMCAdjudicator: WRONG_ARRAY_LENGTHS");
+
     // Verify that the given channel state matches the stored one
     require(hashChannelState(ccs) == channelDispute.channelStateHash, "CMCAdjudicator: INVALID_CHANNEL_HASH");
 
     // We need to be in defund phase for that
     require(inDefundPhase(), "CMCAdjudicator: INVALID_PHASE");
 
-    // We can't defund twice at the same defund nonce
-    // TODO: should this be checked in the `disputeChannel`?
-    require(channelDispute.defundNonce < ccs.defundNonce, "CMCAdjudicator: CHANNEL_ALREADY_DEFUNDED");
-    channelDispute.defundNonce = ccs.defundNonce;
-
     // TODO SECURITY: Beware of reentrancy
     // TODO: offchain-ensure that all arrays have the same length:
-    // assetIds, balances, processedDepositsA, processedDepositsB
+    // assetIds, balances, processedDepositsA, processedDepositsB, defundNonces
     // Make sure there are no duplicates in the assetIds -- duplicates are often a source of double-spends
 
-    // Defund all assets stored in the channel
-    for (uint256 i = 0; i < ccs.assetIds.length; i++) {
-      address assetId = ccs.assetIds[i];
-      Balance memory balance = ccs.balances[i];
+    // Defund all assets given
+    for (uint256 i = 0; i < assetIds.length; i++) {
+      address assetId = assetIds[i];
+
+      // Verify or find the index of the assetId in the ccs.assetIds
+      uint256 index;
+      if (i < indices.length) {
+        // The index was supposedly given -- we verify
+        index = indices[i];
+        require(assetId == ccs.assetIds[index], "CMCAdjudicator: INDEX_MISMATCH");
+      } else {
+        // we search through the assets in ccs
+        for (index = 0; index < ccs.assetIds.length; index++) {
+          if (assetId == ccs.assetIds[index]) {
+            break;
+          }
+        }
```

Comment:
If the asset id is not found in the `ccs.assetIds` there won't be any reverts here, if that's the case then im not sure what exactly the for loop is adding

Reply:
Ahh okay, I see. Correct if wrong:

The index is being assigned as the counter within the for loop, and if it does not exist in the channel state (i.e. a deposit of a new asset was sent to chain, but not reconciled into the channel), then it will be transferred out as an unprocessed deposit

Reply:
Exactly.